File: epi_judge_python/sorted_arrays_merge.py
from test_framework import generic_test
import heapq

def merge_sorted_arrays(sorted_arrays):
    if not sorted_arrays:
        return sorted_arrays

    heap = []
    for i, arr in enumerate(sorted_arrays):
        if len(arr) > 0:
            heap.append((arr.pop(0), i))
    heapq.heapify(heap)

    res = []
    while heap:
        smallest = heapq.heappop(heap)
        corresp_arr = sorted_arrays[smallest[1]]
        if len(corresp_arr) > 0:
            heapq.heappush(heap, (corresp_arr.pop(0), smallest[1]))
        res.append(smallest[0])



    return res


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("sorted_arrays_merge.py",
                                       "sorted_arrays_merge.tsv",
                                       merge_sorted_arrays))

File: epi_judge_java/epi/SortedArraysMerge.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.*;

public class SortedArraysMerge {
  private static class Pair{
    int val;
    Iterator<Integer> iter;

    public Pair(int val, Iterator<Integer>  iter){
      this.val = val;
      this.iter = iter;
    }


  }
  @EpiTest(testDataFile = "sorted_arrays_merge.tsv")

  public static List<Integer>
  mergeSortedArrays(List<List<Integer>> sortedArrays) {
    PriorityQueue<Pair> heap = new PriorityQueue<Pair>(sortedArrays.size(), (Pair a, Pair b)-> Integer.compare(a.val,b.val));
    for (List<Integer> l : sortedArrays){
      if (!l.isEmpty()){
        Iterator<Integer> it = l.iterator();
        Integer v = it.next();
        heap.add(new Pair(v,it));
      }
    }
    List<Integer> res = new ArrayList<>();
    while (!heap.isEmpty()){
      Pair current = heap.poll();
      if (current.iter.hasNext()) {
        heap.add(new Pair(current.iter.next(), current.iter));
      }
      res.add(current.val);
    }
    return res;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SortedArraysMerge.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
